### ISSUE TITLE
`TimeConstraint` inconsistency, `max_duration` wildness

### ISSUE THREAD

**Issue by Tomiinek** [lhotse-speech/lhotse#1145]
I am trying to use a bucketing sampler with a smaller dataset (~10h), and bigger batches (~30min). I want to keep number of buckets low, while enjoying balanced batches with roughly the same duration. However, the bucketing sampler can easily yield batches with only 25% of `max_duration`, which seems to be a bit insane.

The criterion inside `exceeded` of `TimeConstraint` seems to be "memory" driven, and expects that people will use paddings or whatever: https://github.com/lhotse-speech/lhotse/blob/b98521b62bee00bff8a396e991bbec69c4da2f39/lhotse/dataset/sampling/base.py#L349-L356 Would it be possible to add a sampler's option to decide based on `effective_duration > self.max_duration` or `self.duration > self.max_duration`?

Moreover, the criterion inside `close_to_exceeding` of  `TimeConstraint` seems to track the real duration of the batch: https://github.com/lhotse-speech/lhotse/blob/b98521b62bee00bff8a396e991bbec69c4da2f39/lhotse/dataset/sampling/base.py#L365-L372 So I can imagine batches for which `close_to_exceeding` is False but `exceeded` True, isn't it :sweat_smile:

**Comment by pzelasko** [comment]
> I am trying to use a bucketing sampler with a smaller dataset (~10h), and bigger batches (~30min). I want to keep number of buckets low, while enjoying balanced batches with roughly the same duration. However, the bucketing sampler can easily yield batches with only 25% of `max_duration`, which seems to be a bit insane.

That would give you roughly 20 batches per epoch; are you sure there was sufficient amount of data to satisfy `max_duration`? You can enable `drop_last=True`, it's likely the partial batches will disappear.

> Would it be possible to add a sampler's option to decide based on `self.num_cuts * self.longest_seen > self.max_duration` (current state) or `self.duration > self.max_duration`?

That was the original behavior, and both were supported for a time. It made tuning `max_duration` more difficult and was eventually removed in https://github.com/lhotse-speech/lhotse/pull/756 (look for `strict` option).

> Moreover, the criterion inside `close_to_exceeding` of `TimeConstraint` seems to track the real duration of the batch:

Good point -- technically it would only matter when you're using `drop_last=True`, but it's probably desirable to be consistent here.

**Comment by Tomiinek** [comment]
> That would give you roughly 20 batches per epoch; are you sure there was sufficient amount of data to satisfy max_duration? 

Yes I am.

> You can enable drop_last=True, it's likely the partial batches will disappear.

Nono it does not change anything unfortunately ... because the constraint is still what it is :shrug: It does not depend on the size of the buckets. Imagine 100000h of data, 2 buckets, one with data from 1s to 100s, another with data from 100s to 1000s), let max duration be 999s. Sampler starts sampling from the first bucket, and luckily gets 9x 1s samples, then a 100s sample. So the condition is fullfileed (because `100 * 10 > 999`) and you get a batch with duration that is only 11% of max duration. Kind of unexpected ... And this seems to happen very often in scenarios with a low number of batches (e.g. any fine-tuning)

Changing 
```
effective_duration = self.num_cuts * self.longest_seen 
return effective_duration > self.max_duration 
```
to 
```
return self.current > self.max_duration 
```
results in very consistent batches of almost equal size. 

> It made tuning max_duration more difficult and was eventually removed in https://github.com/lhotse-speech/lhotse/pull/756 (look for strict option).

Do you mean tuning to avoid OOM?

**Comment by pzelasko** [comment]
> Do you mean tuning to avoid OOM?

Yeah. What you describe though is a good argument to bring the option back that we haven't thought about before. Would you like to make a PR that brings `strict` option back (enabled by default, ideally with its doc mentioning that it's useful to turn off for fine-tuning on small data)?

**Comment by Tomiinek** [comment]
Yup, I will make the PR :slightly_smiling_face:
